src/server_manager/web_app/digitalocean_app/create_server_app.ts

```diff
@@ -0,0 +1,77 @@
+/*
+  Copyright 2018 The Outline Authors
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+import {customElement, html, LitElement, property} from 'lit-element';
+
+import * as server from '../../model/server';
+import {App, NotificationManager} from '../app';
+
+import {Location, OutlineRegionPicker} from '../ui_components/outline-region-picker-step';
+
+// DigitalOcean mapping of regions to flags
+const FLAG_IMAGE_DIR = 'images/flags';
+const DIGITALOCEAN_FLAG_MAPPING: {[cityId: string]: string} = {
+  ams: `${FLAG_IMAGE_DIR}/netherlands.png`,
+  sgp: `${FLAG_IMAGE_DIR}/singapore.png`,
+  blr: `${FLAG_IMAGE_DIR}/india.png`,
+  fra: `${FLAG_IMAGE_DIR}/germany.png`,
+  lon: `${FLAG_IMAGE_DIR}/uk.png`,
+  sfo: `${FLAG_IMAGE_DIR}/us.png`,
+  tor: `${FLAG_IMAGE_DIR}/canada.png`,
+  nyc: `${FLAG_IMAGE_DIR}/us.png`,
+};
+
+@customElement('digital-ocean-create-server')
+export class DigitalOceanCreateServer extends LitElement {
+  @property({type: Function}) localize: Function;
+  @property({type: Object}) app: App = null;
+  @property({type: Object}) digitalOceanRepository: server.ManagedServerRepository = null;
+  @property({type: Object}) notificationManager: NotificationManager = null;
+
+  render() {
+    return html`
+        <outline-region-picker-step id="regionPicker" .localize=${
+        this.localize}></outline-region-picker-step>
+    `;
+  }
+
+  private createLocationModel(cityId: string, regionIds: string[]): Location {
+    return {
+      id: regionIds.length > 0 ? regionIds[0] : null,
+      name: this.localize(`city-${cityId}`),
+      flag: DIGITALOCEAN_FLAG_MAPPING[cityId] || '',
+      available: regionIds.length > 0,
+    };
+  }
+
+  // The region picker initially shows all options as disabled. Options are enabled by this code,
+  // after checking which regions are available.
+  async show() {
+    const regionPicker = this.shadowRoot.querySelector('#regionPicker') as OutlineRegionPicker;
+    regionPicker.reset();
+
+    try {
+      const map =
+          await this.app.digitalOceanRetry(() => this.digitalOceanRepository.getRegionMap());
```

Comment:
Can you remove the dependency on app , since all you need is `digitalOceanRetry`?

Reply:
Absolutely.

I updated the code to pass in the retry function, but I'd like to make that function more generic in the future. 

I moved the digital ocean specific error handling into the DO API, but the retry function also has a UI. We should be able to reuse the same connectivity error dialog for other cloud providers once we make the underlying UI business logic generic.

<img width="573" alt="Screen Shot 2020-09-18 at 9 53 32 AM" src="https://user-images.githubusercontent.com/1009390/93607672-a4e3ec80-f997-11ea-9841-ef1b8629cf08.png">